File: implementations/vb/Loom.Core/Models/ValidationResult.vb
Imports Loom.Core.Interfaces

Namespace Models
    ''' <summary>
    ''' Implements IValidationResult
    ''' </summary>
    Public Class ValidationResult
        Implements Interfaces.IValidationResult

        Private ReadOnly _isValid As Boolean
        Private ReadOnly _errors As New List(Of String)

        Public ReadOnly Property IsValid As Boolean Implements IValidationResult.IsValid
            Get
                Return _isValid
            End Get
        End Property

        Public ReadOnly Property Errors As IEnumerable(Of String) Implements IValidationResult.Errors
            Get
                Return _errors
            End Get
        End Property

        ''' <summary>
        ''' Validator constructor
        ''' </summary>
        ''' <param name="isValid"> Boolean </param>
        ''' <param name="errors"> List of String </param>
        Public Sub New(isValid As Boolean, Optional errors As List(Of String) = Nothing)
            _isValid = isValid
            If errors IsNot Nothing Then _errors.AddRange(errors)
        End Sub

        ''' <summary>
        ''' Success
        ''' </summary>
        ''' <returns> No error (True) </returns>
        Public Shared Function Success() As ValidationResult
            Return New ValidationResult(True)
        End Function

        ''' <summary>
        ''' Failure
        ''' </summary>
        ''' <param name="errors"> List of String</param>
        ''' <returns> False + Errors </returns>
        Public Shared Function Failure(ParamArray errors As String()) As ValidationResult
            Return New ValidationResult(False, errors.ToList())
        End Function

    End Class
End Namespace

File: implementations/csharp/ValidationResult.cs
using Loom.Core.Interfaces;
using System.Collections.Generic;
using System.Linq;

namespace Models
{
    /// <summary>
    /// Implements IValidationResult
    /// </summary>
    public class ValidationResult : Interfaces.IValidationResult
    {
        private readonly bool _isValid;
        private readonly List<string> _errors = new List<string>();

        public bool IsValid => _isValid;
        public IEnumerable<string> Erros => _errors;

        /// <summary>
        /// Validator constructor
        /// </summary>
        /// <param name="isValid"> Boolean </param>
        /// <param name="errors"> List of String </param>
        public ValidationResult(bool isValid, List<string> errors = null)
        {
            _isValid = isValid;
            if (errors != null)
            {
                _errors.AddRange(errors);
            }
        }

        /// <summary>
        /// Success
        /// </summary>
        /// <returns> No error (True) </returns>
        public static ValidationResult Success()
        {
            return new ValidationResult(true);
        }

        /// <summary>
        /// Failure
        /// </summary>
        /// <param name="errors"> List of String </param>
        /// <returns> False + Errors </returns>
        public static ValidationResult Failure(params string[] errors)
        {
            return new ValidationResult(false, errors.ToList());
        }
    }
}